### PR TITLE
DLAF_AddTest improvement

### LINKED ISSUE #43: Specify the main routine for tests.
`DLAF_addTest` should have a parameter to specify to which main function to link:

- NO -> custom main
- PLAIN -> main included in gtest
- HPX -> main starts and stops HPX
- MPI -> main initializes and finalizes MPI
- MPIHPX -> main initializes MPI and HPX

and others when needed.

### PR DESCRIPTION
**Pull request by albestro**
Close #43 

- new option MPI_OVERSUBSCRIBE
- changed option from USE_GTEST_MAIN to USE_MAIN, which allows to
select different types of external main (PLAIN, HPX, MPI, MPIHPX)
- update related documentation

### CODE REVIEW

File: cmake/DLAF_AddTest.cmake [hunk 1]
```diff
@@ -67,8 +88,19 @@ function(DLAF_addTest test_target_name)
...
+
+  if (DLAF_AT_USE_MAIN)
+    if (DLAF_AT_USE_MAIN STREQUAL PLAIN)
+      target_link_libraries(${test_target_name} PRIVATE gtest_main)
+    elseif (DLAF_AT_USE_MAIN STREQUAL HPX)
+      target_link_libraries(${test_target_name} PRIVATE DLAF_gtest_hpx_main)
+    elseif (DLAF_AT_USE_MAIN STREQUAL MPI)
+      target_link_libraries(${test_target_name} PRIVATE DLAF_gtest_mpi_main)
+    elseif (DLAF_AT_USE_MAIN STREQUAL MPIHPX)
+      message(FATAL_ERROR "USE_MAIN=${DLAF_AT_USE_MAIN} is not yet supported")
```
Comment (albestro) [inline cmake/DLAF_AddTest.cmake]: @rasolca, I have implemented the external main for MPI but not for the combo MPIHPX.

File: cmake/DLAF_AddTest.cmake [hunk 2]
```diff
@@ -86,10 +118,14 @@ function(DLAF_addTest test_target_name)
       message(FATAL_ERROR "Wrong MPIRANKS number ${DLAF_AT_MPIRANKS}")
     endif()
 
-    if (DLAF_AT_MPIRANKS GREATER MPIEXEC_MAX_NUMPROCS)
-      message(FATAL_ERROR "Impossible to have more than ${MPIEXEC_MAX_NUMPROCS}")
+    if (DLAF_AT_MPIRANKS GREATER MPIEXEC_MAX_NUMPROCS AND NOT DLAF_AT_MPI_OVERSUBSCRIBE)
```
Comment (rasolca) [inline cmake/DLAF_AddTest.cmake]: `max_numprocs` can be set to any number. How can you know if you need `OVERSUBSCRIBE`?

File: cmake/DLAF_AddTest.cmake [hunk 3]
```diff
@@ -118,10 +110,11 @@ function(DLAF_addTest test_target_name)
       message(FATAL_ERROR "Wrong MPIRANKS number ${DLAF_AT_MPIRANKS}")
     endif()
 
-    if (DLAF_AT_MPIRANKS GREATER MPIEXEC_MAX_NUMPROCS AND NOT DLAF_AT_MPI_OVERSUBSCRIBE)
+    if (DLAF_AT_MPIRANKS GREATER MPIEXEC_MAX_NUMPROCS)
       message(WARNING "\
         YOU ARE ASKING FOR ${DLAF_AT_MPIRANKS} RANKS, BUT THERE ARE JUST ${MPIEXEC_MAX_NUMPROCS} CORES.
-        Use MPI_OVERSUBSCRIBE flag to overcome this: be careful, since this may lead to deadlocks.")
+        You can adjust MPIEXEC_MAX_NUMPROCS value to overcome this.
```
Comment (rasolca) [inline cmake/DLAF_AddTest.cmake]: I would change the last part of the sentence to "to suppress this warning."

File: cmake/DLAF_AddTest.cmake [hunk 4]
```diff
@@ -118,10 +110,11 @@ function(DLAF_addTest test_target_name)
...
+        With OpenMPI it may be needed to set environment variable OMPI_MCA_rmaps_base_oversubscribe=1.")
```
Comment (rasolca) [inline cmake/DLAF_AddTest.cmake]: ... set **the** environment ...

**Pull request merged**